### ISSUE TITLE
x86 vs x64 APIENTRY?

### ISSUE THREAD

**Issue by JohannesKauffmann** [lu-zero/mfx_dispatch#88]
Compiling for x86/i686 UWP/Windows Store. Library which is statically linked fails to load `intel_gfx_api-x86.dll`, with Entry Point Not Found. This is because `intel_gfx_api-x86.dll` (on my machine) exports `_InitialiseMediaSession` and `_DisposeMediaSession`, but `libmfx.a` imports `_InitialiseMediaSession@12` and `_DisposeMediaSession@4`.

It works fine on x64, since the `APIENTRY` modifiers in `intel_api_factory.h` are ignored on x64.

Now I managed to hack around this by patching `intel_gfx_api-x86.def`, removing the name decorations, and by patching `intel_api_factory.h` by removing the `APIENTRY` modifiers.

But would that be the right solution? Is it an Intel problem that the x86 dll exports non-decorated function names (meaning no @12 or @4 appended)? Should this library be modified to match the dll? Or should Intel update their dll's? Looking through other x86 system dll's, it seems none of them have these name decorations in their export table.

My machine with Intel Xe Graphics, and my buddies machine with an Intel UHD630, both have the x86 dll which exports the non-decorated names.

Edit: this is what I changed to make it work:
```
diff --git a/src/intel_api_factory.h b/src/intel_api_factory.h
index 0be5dcb..29e9773 100644
--- a/src/intel_api_factory.h
+++ b/src/intel_api_factory.h
@@ -25,8 +25,8 @@ extern "C"
 {
 #endif /* __cplusplus */

-HRESULT APIENTRY InitialiseMediaSession(_Out_ HANDLE* handle, _In_ LPVOID lpParam, _Reserved_ LPVOID lpReserved);
-HRESULT APIENTRY DisposeMediaSession(_In_ const HANDLE handle);
+HRESULT InitialiseMediaSession(_Out_ HANDLE* handle, _In_ LPVOID lpParam, _Reserved_ LPVOID lpReserved);
+HRESULT DisposeMediaSession(_In_ const HANDLE handle);

 #ifdef __cplusplus
 }
diff --git a/src/intel_gfx_api-x86.def b/src/intel_gfx_api-x86.def
index 19d0727..c9c15e6 100644
--- a/src/intel_gfx_api-x86.def
+++ b/src/intel_gfx_api-x86.def
@@ -1,4 +1,4 @@
 LIBRARY intel_gfx_api-x86.dll
 EXPORTS
-InitialiseMediaSession@12
-DisposeMediaSession@4
+InitialiseMediaSession^M
+DisposeMediaSession^M
```

**Comment by JohannesKauffmann** [comment]
See also here: https://code.videolan.org/videolan/LibVLCSharp/-/issues/374

**Comment by lu-zero** [comment]
I'm waiting for feedback I'd accept it if Intel upstream accepts the change.

**Comment by JohannesKauffmann** [comment]
You mean, I should first go to [Intel MediaSDK](https://github.com/Intel-Media-SDK/MediaSDK), ask them to remove the `APIENTRY` modifier and report back? Or have you contacted them and are waiting for feedback from them?

In any case, I don't think they will agree to remove the `APIENTRY` modifier, since removing it could result in all kinds of unforeseen breakage on x86. In fact, using my workaround shown above, the library will link correctly, but I haven't tested at all if it doesn't break basic functionality.

Intel doesn't need to remove the modifier from their code, since they get a handle to the `InitialiseMediaSession` and `DisposeMediaSession` using [`GetProcAddr`](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/api/mfx_dispatch/windows/src/mfx_dispatcher_uwp.cpp#L49), which allows for spelling the function name correctly in the first place. Only after they've aquired a handle to the function, this [handle is casted](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/api/mfx_dispatch/windows/src/mfx_dispatcher_uwp.cpp#L58) to a function pointer type with the `APIENTRY` modifier.

Intel's upstream mfx_dispatch loads the `intel_gfx_api.dll` dynamically at runtime using `LoadLibraryExW`, but `libmfx.a` links statically to `intel_gfx_api.dll` using function prototypes which include the `APIENTRY` modifier, which means it won't link correctly on x86.

Maybe this could be solved in the same way Intel does it. Link statically to the correct function name (without the modifer), and during runtime, cast the symbol or function pointer to a type which includes the modifier (is this even possible)?

**Comment by maximd33** [comment]
if change above works , it should be fine to be merged and no objections should be expected.

MSDK uses [dynamic way of working with InitialiseMediaSession](https://github.com/Intel-Media-SDK/MediaSDK/blob/510d19dcace1d8c57567fdd40b557155ab11ab8e/api/mfx_dispatch/windows/src/mfx_dispatcher_uwp.cpp)....
so global rework can be expected but only later.

**Comment by JohannesKauffmann** [comment]
I would, however, I just realized that I have been building and testing with a different version than the current master...

Version from commit 7efc7505465bc1f16fbd1da3d24aa5bd9d46c5ca is used in VLC, along with a patch. [See here](https://code.videolan.org/videolan/vlc/-/blob/3.0.x/contrib/src/mfx/rules.mak). This version directly links with `intel_gfx_api.dll`, while master includes files from IntelMediaSDK upstream (resulting in a build error for mater). It seems like loading the library is offloaded to this Intel code, which loads this lib at runtime.

Is it the intention to later bring back the implementation from version 7efc7505465bc1f16fbd1da3d24aa5bd9d46c5ca and earlier, which linked directly with `intel_gfx_api.dll`? As far as I understand, loading the library at runtime using [`LoadLibraryExW`](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/api/mfx_dispatch/windows/src/mfx_load_dll.cpp#L161) fails on UWP.

To summarize, my fix only really works for older versions of this mfx_dispatch lib. Maybe @robUx4 could also take a look at this, since he originally implemented UWP support for mfx_dispatch, and it concerns VLC?

**Comment by robUx4** [comment]
On UWP we can't use `LoadLibraryExW` for DLL outside of the app, nor read the registry. VLC got a special exception to be allowed to use these particular DLLs from Intel for UWP. The code update lost proper UWP support and that's a reason why I didn't update the code in VLC (yet).

**Comment by JohannesKauffmann** [comment]
Right, my patch (above) was for the version VLC is using right now (7efc7505465bc1f16fbd1da3d24aa5bd9d46c5ca), since it links to the wrong symbols on x86, which is caused by the APIENTRY modifier. Should I then propose the patch to VLC, or here in case the UWP code comes back and sumbit a backport patch in the meantime to VLC?
